src/mappings/horizonStaking.ts

```diff
@@ -0,0 +1,78 @@
+import { BigInt } from '@graphprotocol/graph-ts'
+import { addresses } from '../../config/addresses.template'
+import { HorizonStakeDeposited, HorizonStakeLocked, HorizonStakeWithdrawn} from '../types/HorizonStaking/HorizonStaking'
+import { Indexer } from '../types/schema'
+import { calculateCapacities, createOrLoadEpoch, createOrLoadGraphNetwork, createOrLoadIndexer, updateAdvancedIndexerMetrics } from './helpers/helpers'
+
+
+export function handleHorizonStakeDeposited(event: HorizonStakeDeposited): void {
+    // To Do
+    let graphNetwork = createOrLoadGraphNetwork(event.block.number, event.address)
+    // update indexer
+    let indexer = createOrLoadIndexer(event.params.serviceProvider, event.block.timestamp)
+    let previousStake = indexer.stakedTokens
+    indexer.stakedTokens = indexer.stakedTokens.plus(event.params.tokens)
+    indexer = updateAdvancedIndexerMetrics(indexer as Indexer)
+    indexer = calculateCapacities(indexer as Indexer)
+    indexer.save()
+  
+    // Update graph network
+    graphNetwork.totalTokensStaked = graphNetwork.totalTokensStaked.plus(event.params.tokens)
+    if (previousStake == BigInt.fromI32(0)) {
+      graphNetwork.stakedIndexersCount = graphNetwork.stakedIndexersCount + 1
+    }
+    graphNetwork.save()
+  
+    // Update epoch
+    let epoch = createOrLoadEpoch(
+      addresses.isL1 ? event.block.number : graphNetwork.currentL1BlockNumber!,
+    )
+    epoch.stakeDeposited = epoch.stakeDeposited.plus(event.params.tokens)
+    epoch.save()
+}
+
+export function handleHorizonStakeLocked(event: HorizonStakeLocked): void {
+  // To Do
+  let graphNetwork = createOrLoadGraphNetwork(event.block.number, event.address)
+  // update indexer
+  let id = event.params.serviceProvider.toHexString()
+  let indexer = Indexer.load(id)!
+  let oldLockedTokens = indexer.lockedTokens
+  indexer.lockedTokens = event.params.tokens
+  indexer.tokensLockedUntil = event.params.until.toI32()
+  indexer = updateAdvancedIndexerMetrics(indexer as Indexer)
+  indexer = calculateCapacities(indexer as Indexer)
+  indexer.save()
+
+  // update graph network
+  // the tokens from the event replace the previously locked tokens
+  // from this indexer
+  graphNetwork.totalUnstakedTokensLocked = graphNetwork.totalUnstakedTokensLocked.plus(
+    event.params.tokens,
+  ).minus(oldLockedTokens)
```

Comment:
Is this `minus(oldLockedTokens)` correct? The value in `event.params.tokens` does not include `oldLockedTokens` I think, so don't think we should substract? Unless I'm missing something.

Reply:
This is how it was implemented in V1, if I'm not mistaken that's to take into consideration the forced release/withdrawal of funds that could be locked from a previous StakeLocked, but taking a closer look at this it does sound like it could actually be quite buggy. (as that withdrawal only happens if the previously locked tokens are actually withdrawable)